google/cloud/status.h

```diff
@@ -82,42 +82,40 @@ absl::optional<std::string> GetPayload(Status const&, std::string const& key);
 class Status {
  public:
   Status() = default;
+  ~Status();
+  Status(Status const&);
+  Status& operator=(Status const&);
+  Status(Status&&) noexcept;
+  Status& operator=(Status&&) noexcept;
 
   /**
    * Constructs a Status with the given @p code and @p message.
    *
    * Ignores @p message if @p code is `StatusCode::kOk`.
    */
-  explicit Status(StatusCode code, std::string message)
-      : code_(code),
-        message_(code_ == StatusCode::kOk ? "" : std::move(message)) {}
-
-  bool ok() const { return code_ == StatusCode::kOk; }
-  StatusCode code() const { return code_; }
-  std::string const& message() const { return message_; }
-
-  friend inline bool operator==(Status const& a, Status const& b) {
-    return a.code_ == b.code_ && a.message_ == b.message_ &&
-           a.payload_ == b.payload_;
-  }
-  friend inline bool operator!=(Status const& a, Status const& b) {
-    return !(a == b);
-  }
-  friend inline std::ostream& operator<<(std::ostream& os, Status const& s) {
-    if (s.ok()) return os << s.code();
-    return os << s.code() << ": " << s.message();
-  }
+  explicit Status(StatusCode code, std::string message);
+
+  bool ok() const { return impl_ == nullptr; }
+  StatusCode code() const;
+  std::string const& message() const;
+
+  friend bool operator==(Status const& a, Status const& b);
+  friend bool operator!=(Status const& a, Status const& b);
+  friend std::ostream& operator<<(std::ostream& os, Status const& s);
 
  private:
   friend void internal::SetPayload(Status&, std::string, std::string);
   friend absl::optional<std::string> internal::GetPayload(Status const&,
                                                           std::string const&);
 
-  StatusCode code_{StatusCode::kOk};
-  std::string message_;
-  std::unordered_map<std::string, std::string> payload_;
+  class Impl;
+  // A null `impl_` is an OK status. Only non-OK Statuses allocate an Impl.
+  Impl* impl_{nullptr};
```

Comment:
Why not use `std::unique_ptr<>` and keep `~Status()` in the `.cc` file?

Reply:
std::unique_ptr requires a `Impl` to be fully defined, which would mean the definition of Impl needs to be pulled out into som other header. `std::shared_ptr` allow (I think) a forward decl of `Impl` only, so it would work, but shared_ptr implements "sharing" semantics which we don't need. But I think I could switch to shared_ptr still w/ a custom implementation of the copy c'tor and assignment to implement deep copy.

Reply:
@coryan pointed out that I was mistaken, and `std::unique_ptr` does work here. Changed. Looks better to me.